### PR TITLE
std: Stabilize APIs for the 1.6 release

### LINKED ISSUE #27585: Tracking issue for `read_exact` (RFC #980)
https://github.com/rust-lang/rfcs/pull/980

### LINKED ISSUE #27704: Tracking issue for OsStr conversions
This is a tracking issue for stabilizing the `convert` feature in the standard library, which is primarily associated with converting `OsStr` and `OsString` to byte or string types. Due to the length of discussions on #27390 and #27657 this will likely required an RFC to move forward (I hope to write one soon)

### LINKED ISSUE #27707: Tracking issue for fs::walk_dir
This is a tracking issue for the `fs_walk` unstable feature in the standard library. Some of the open questions currently are:

* There are many many ways to walk a directory, and there probably wants to be something like a `WalkBuilder` structure.
* The current walking strategy holds open a lot of file descriptors - #23715
* The default walking strategy has not been well audited for all manner of fs corner cases

The [C++ `<filesystem>` implementation](http://en.cppreference.com/w/cpp/experimental/fs/recursive_directory_iterator) of walking directories is quite a useful reference in terms of what options we may want as well as what we haven't implemented. Note that this is a large enough feature that this may want to be prototyped outside the standard library first and then perhaps consider moving in at a later date.

### LINKED ISSUE #27710: Tracking issue for fs::DirBuilder
This is a tracking issue for the unstable `dir_builder` feature in the standard library. The current purpose of this is to provide access to platform-specific options such as modes on Unix or security attributes on Windows. The current `create_dir` and `create_dir_all` APIs are built on top of this API right now.

There aren't too many open questions other than bikeshedding things like naming and such. [C++ may have some inspiration](http://en.cppreference.com/w/cpp/experimental/fs) as well for an API like this.

### LINKED ISSUE #27711: Tracking issue for `drain` stabilization
The `drain` method was originally proposed in [collections reform 2](https://github.com/rust-lang/rfcs/pull/509) and then adjusted in a [follow-up RFC](https://github.com/rust-lang/rfcs/pull/574). The main remaining question for the API is the traits it uses to accept various kinds of ranges, which potentially interacts with [inclusive ranges](https://github.com/rust-lang/rfcs/pull/1192).

cc @bluss, who is working on an RFC to finish off this API.

### LINKED ISSUE #27727: Tracking issue for `path_components_peek` stabilization
The `path::Components::peek` method allows you to peek at the next component in the iterator without actually consuming it. There's no strong reason for it to remain unstable, except that the conventions around such methods are a bit unclear.

### LINKED ISSUE #27740: Tracking issue for std::slice::bytes
This is a tracking issue for the unstable `slice_bytes` feature in the standard library. This module has some various helper methods for dealing with blocks of bytes in an efficient fashion, but they seem particularly one-off and not necessarily fitting in with the rest of the design of the standard library.

Some drawbacks are:

* A one-off trait `MutableByteVector` is used to give slices the `set_memory` method.
* A one-off and oddly placed `copy_memory` function is used to provide essentially safe bindings to `memcpy`

Both of these can be implemented in stable Rust otherwise and get optimized down to the same implementation, but it's often useful to call them directly to guarantee the optimization will happen. Needs a decision to whether this functionality should be exposed, where it should be exposed, and how.

### LINKED ISSUE #27744: Tracking issue for `push_all` stabilization
The `Vec::push_all` method special-cases pushing a slice onto a vector, and historically has performed much better than using `extend` (depending on how LLVM optimizations are working out).

We need to offer a method with maximal performance somehow, but ideally this would come from `extend`, either through optimization or through [impl specialization](https://github.com/rust-lang/rfcs/pull/1210).

### LINKED ISSUE #27799: Tracking issue for Duration::span
This is a tracking issue for the unstable `duration_span` feature of the standard library. Known open questions are:

* Is this the right place for this API?
* Is this the right name?
* Should the closure be able to return a value?
* Should various other system clocks be exposed?

### LINKED ISSUE #27801: Tracking issue for IpAddr
This is a tracking issue for the unstable `ip_addr` feature of the standard library. Some thoughts:

* This isn't actually ever used in the system APIs as there's no notion of "an IP address", just an IPv4 or IPv6 address.
* We have no stable APIs producing this value.
* Some old UDP APIs took this, but they didn't actually want to
* The DNS APIs may want to produce or consume this.

### LINKED ISSUE #28968: Tracking Issue for Mutex and RWLock unique accessors
After #29031, this is now the tracking issue for the stabilization of the following features:
 * [ ] mutex_into_inner
 * [ ] mutex_get_mut
 * [ ] rwlock_into_inner
 * [ ] rwlock_get_mut

## Original report

As far as I can tell, it's sound to straight-up-unwrap a Mutex if you have it by value. This would enable one to do:

```rust
let data = Mutex::new(vec![])
crossbeam::scope(|scope| {
  for input in inputs {
    let data = &data;
    scope.spawn(move || {
      let output = expensive_computation(input);
      data.lock().unwrap().push(output);
    }
  }
});

for expensive in data.into_inner() {
  non_thread_safe(expensive);
}
```

Which today requires mutexing an `&mut`:

```rust
let mut data = vec![];
{
  let data_mutex = Mutex::new(&mut data);
  crossbeam::scope(|scope| {
    for input in inputs {
      let data = &data_mutex;
      scope.spawn(move || {
        let output = expensive_computation(input);
        data.lock().unwrap().push(output);
      }
    }
  });
}

for expensive in data.into_inner() {
  non_thread_safe(expensive);
}
```

Given Arc can be unwrapped, this would also allow the last `Arc<Mutex<T>>` in some concurrent context to be fully unwrapped to a `T`.

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is the standard API stabilization commit for the 1.6 release cycle.
The list of issues and APIs below have all been through their cycle-long FCP and
the libs team decisions are listed below

Stabilized APIs
- `Read::read_exact`
- `ErrorKind::UnexpectedEof` (renamed from `UnexpectedEOF`)
- libcore -- this was a bit of a nuanced stabilization, the crate itself is now
  marked as `#[stable]` and the methods appearing via traits for primitives like
  `char` and `str` are now also marked as stable. Note that the extension traits
  themeselves are marked as unstable as they're imported via the prelude. The
  `try!` macro was also moved from the standard library into libcore to have the
  same interface. Otherwise the functions all have copied stability from the
  standard library now.
- `fs::DirBuilder`
- `fs::DirBuilder::new`
- `fs::DirBuilder::recursive`
- `fs::DirBuilder::create`
- `os::unix::fs::DirBuilderExt`
- `os::unix::fs::DirBuilderExt::mode`
- `vec::Drain`
- `vec::Vec::drain`
- `string::Drain`
- `string::String::drain`
- `vec_deque::Drain`
- `vec_deque::VecDeque::drain`
- `collections::hash_map::Drain`
- `collections::hash_map::HashMap::drain`
- `collections::hash_set::Drain`
- `collections::hash_set::HashSet::drain`
- `collections::binary_heap::Drain`
- `collections::binary_heap::BinaryHeap::drain`
- `Vec::extend_from_slice` (renamed from `push_all`)
- `Mutex::get_mut`
- `Mutex::into_inner`
- `RwLock::get_mut`
- `RwLock::into_inner`
- `Iterator::min_by_key` (renamed from `min_by`)
- `Iterator::max_by_key` (renamed from `max_by`)

Deprecated APIs
- `ErrorKind::UnexpectedEOF` (renamed to `UnexpectedEof`)
- `OsString::from_bytes`
- `OsStr::to_cstring`
- `OsStr::to_bytes`
- `fs::walk_dir` and `fs::WalkDir`
- `path::Components::peek`
- `slice::bytes::MutableByteVector`
- `slice::bytes::copy_memory`
- `Vec::push_all` (renamed to `extend_from_slice`)
- `Duration::span`
- `IpAddr`
- `SocketAddr::ip`
- `Read::tee`
- `io::Tee`
- `Write::broadcast`
- `io::Broadcast`
- `Iterator::min_by` (renamed to `min_by_key`)
- `Iterator::max_by` (renamed to `max_by_key`)
- `net::lookup_addr`

New APIs (still unstable)
- `<[T]>::sort_by_key` (added to mirror `min_by_key`)

Closes #27585
Closes #27704
Closes #27707
Closes #27710
Closes #27711
Closes #27727
Closes #27740
Closes #27744
Closes #27799
Closes #27801
cc #27801 (doesn't close as `Chars` is still unstable)
Closes #28968

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @aturon 

cc @rust-lang/libs

**Comment by alexcrichton** [comment]
This commit also notably stabilizes `no_std` as-is. It didn't seem like there was enough thrust and consensus on https://github.com/rust-lang/rust/issues/27701 for renaming .

File: src/liballoc/boxed.rs [hunk 1]
```diff
@@ -527,6 +526,7 @@ impl<I: ExactSizeIterator + ?Sized> ExactSizeIterator for Box<I> {}
 /// ```
 #[rustc_paren_sugar]
 #[unstable(feature = "fnbox", reason = "Newly introduced", issue = "0")]
+#[fundamental]
```
Comment (aturon) [inline src/liballoc/boxed.rs]: Just slip that in there :)

This is probably prudent, and technically breaking but it's unstable anyway.

File: src/libcore/slice.rs [hunk 2]
```diff
@@ -61,72 +61,98 @@ use raw::Slice as RawSlice;
 //
 
 /// Extension methods for slices.
-#[allow(missing_docs)] // docs in libcollections
-#[doc(hidden)]
 #[unstable(feature = "core_slice_ext",
            reason = "stable interface provided by `impl [T]` in later crates",
            issue = "27701")]
+#[allow(missing_docs)] // documented elsewhere
```
Comment (aturon) [inline src/libcore/slice.rs]: Perhaps we should add docs now? The duplication is awful, though...

**Comment by aturon** [comment]
This LGTM. We should figure out a story for the libcore docs eventually, but it shouldn't block this PR.

File: src/liballoc/boxed.rs (hunk 1 above)
Comment (alexcrichton) [inline src/liballoc/boxed.rs]: Oh oops, this wasn't supposed to sneak in here

File: src/libcore/slice.rs (hunk 2 above)
Comment (alexcrichton) [inline src/libcore/slice.rs]: Good point! I meant to open an issue, and [now I have](https://github.com/rust-lang/rust/issues/30214).

**Pull request merged**